examples/reduce.c

```diff
@@ -0,0 +1,364 @@
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s", msg);
+
+		/* we copy the start of me0 to the start of md */
+		md.start = me0.start;
+
+		ret = PtlMDBind(nih, &md, &mdh);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlMDBind failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+
+		ret = PtlTriggeredAtomic(mdh, 0, ranks[rank].size, PTL_ACK_REQ, id, pti2, 0, 0,
+					 NULL, 0, PTL_SUM, PTL_UINT32_T, ranks[rank].counter, 2);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlAtomic failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+
+		ret = PtlEQWait(eqh, &ev);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s", msg);
+
+		ret = PtlEQWait(eqh, &ev);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s", msg);
+
+		ret = PtlEQWait(eqh, &ev);
+		if (ret != PTL_OK) {
+			fprintf(stderr, "PtlEQWait failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+			return 1;
+		}
+		PtlEvToStr(0, &ev, msg);
+		printf("Event : %s", msg);
+
+		int *num0 = me1.start;
+		printf("content of me1 : {%d,%d,%d} \n", num0[0], num0[1], num0[2]);
+	}
+	return 0;
+}
+
+int main(void)
+{
+	int ret;
+	ptl_handle_ni_t nih;
+	ptl_handle_eq_t eqh;
+	ptl_index_t pti0;
+	ptl_index_t pti1;
+	ptl_index_t pti2;
+	ptl_process_t id;
+	ptl_handle_ct_t counter0;
+	ptl_handle_ct_t counter1;
+
+	int rank0[3] = { 1, 2, 3 };
+	int rank1[3] = { 4, 5, 6 };
+	int rank2[3] = { 7, 8, 9 };
+	int rank3[3] = { 10, 11, 12 };
+
+	int data0[3] = { 0, 0, 0 };
+	int data1[3] = { 0, 0, 0 };
+	int data2[3] = { 0, 0, 0 };
+
+	/* init interface and handles */
+	ret = PtlInit();
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlInit failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+
+	ret = PtlNIInit(PTL_IFACE_DEFAULT, PTL_NI_NO_MATCHING | PTL_NI_PHYSICAL, PTL_PID_ANY, NULL,
+			NULL, &nih);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlNIInit failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+
+	ret = PtlEQAlloc(nih, 10, &eqh);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlEQAlloc failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+
+	ret = PtlPTAlloc(nih, PTL_PT_FLOWCTRL, eqh, PTL_PT_ANY, &pti0);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlPTAlloc failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+
+	ret = PtlPTAlloc(nih, PTL_PT_FLOWCTRL, eqh, PTL_PT_ANY, &pti1);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlPTAlloc failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}
+
+	ret = PtlPTAlloc(nih, PTL_PT_FLOWCTRL, eqh, PTL_PT_ANY, &pti2);
+	if (ret != PTL_OK) {
+		fprintf(stderr, "PtlPTAlloc failed : %s \n", PtlToStr(ret, PTL_STR_ERROR));
+		return 1;
+	}

```

Comment:
Those allocations should be performed in `rank_main`. We should also use a different EQ for each rank (allocated in rank_main).

The `main` function should only bother itself with the NIInitialisation.
As we keep talking about `ranks` we _could_ try to perform a PtlSetMap with 4 times the interface, but I'm not sure swptl is going to handle this well